File: data-stream-as-disjoint-intervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class SummaryRanges {
public:
    /** Initialize your data structure here. */
    SummaryRanges() {
        
    }
    
    void addNum(int val) {
        vector<Interval> tmp;
        Interval cur(val, val);
        int index = 0;
        for (int i = 0; i < res.size(); ++i) {
            if (cur.end + 1 == res[i].start) {
                cur.end = res[i].end;
            } else if (cur.start -1 == res[i].end) {
                cur.start = res[i].start;
            } else if (cur.end + 1 < res[i].start) {
                tmp.push_back(res[i]);
            } else if (cur.start - 1 > res[i].end) {
                tmp.push_back(res[i]);
                ++index;
            } else {
                cur.start = min(cur.start, res[i].start);
                cur.end = max(cur.end, res[i].end);
            }
        }
        tmp.insert(tmp.begin() + index, cur);
        res.clear();
        res = tmp;
    }
    
    vector<Interval> getIntervals() {
        return res;
    }
    private:
    vector<Interval> res;
};

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges obj = new SummaryRanges();
 * obj.addNum(val);
 * vector<Interval> param_2 = obj.getIntervals();
 */

File: data-stream-as-disjoint-intervals.java
import java.util.ArrayList;
import java.util.List;

/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class SummaryRanges {

    private List<Interval> res = null;
    /** Initialize your data structure here. */
    public SummaryRanges() {
        res = new ArrayList<Interval>();
    }

    public void addNum(int val) {
        List<Interval> preN = new ArrayList<Interval>();
        List<Interval> postN = new ArrayList<Interval>();
        Interval cur = new Interval(val, val);
        boolean tag = false;
        for (Interval interval : res) {
            if (cur.end + 1 == interval.start) {
                cur.end = interval.end;
            } else if (cur.start -1 == interval.end) {
                cur.start = interval.start;
            } else if (cur.end + 1 < interval.start) {
                postN.add(interval);
            } else if (cur.start - 1 > interval.end) {
                preN.add(interval);
            } else {
                //Keep it in its original state
                //release heap
                preN.clear();
                postN.clear();
                cur = null;
                return;
            }
        }
        List<Interval> tmp = new ArrayList<Interval>();
        for (Interval interval : preN) {
            tmp.add(interval);
        }
        if (!tag) {
            tmp.add(cur);
        }
        for (Interval interval : postN) {
            tmp.add(interval);
        }
        //release heap
        preN.clear();
        postN.clear();
        this.res.clear();
        cur = null;
        this.res = tmp;
    }

    public List<Interval> getIntervals() {
        return res;
    }

    // class Interval {
    //     int start;
    //     int end;

    //     Interval() {
    //         start = 0;
    //         end = 0;
    //     }

    //     Interval(int s, int e) {
    //         start = s;
    //         end = e;
    //     }
    // }

    public void main() {
        SummaryRanges obj = new SummaryRanges();
        obj.addNum(1);
        print(obj);
        obj.addNum(3);
        print(obj);

        obj.addNum(7);
        print(obj);
        obj.addNum(2);
        print(obj);
        obj.addNum(6);
        print(obj);
    }
    public void print(SummaryRanges obj) {
        List<Interval> param_2 = obj.getIntervals();
        for (Interval t : param_2) {
            System.out.println(t.start + " " + t.end);
        }
        System.out.println("===========");
    }

    public static void main(String[] args) {
        new SummaryRanges().main();
    }
}

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges obj = new SummaryRanges();
 * obj.addNum(val);
 * List<Interval> param_2 = obj.getIntervals();
 */